### ISSUE TITLE
Refactor type imports to one package; make fake imports consistent

### ISSUE THREAD

**Issue by tjcouch-sil** [paranext/paranext-core#650]
**User Story**
As an extension developer, I want consistent and understandable imports so that I can find what I want to use and it doesn't break when the core team moves files around.

**Description**
We should refactor the way we expose types to extensions. Right now, type imports are tied to our file structure like so:
```typescript
import type { ExecutionActivationContext } from 'extension-host/extension-types/extension-activation-context.model';
```

We need to change this. We can probably create a new type export file similar to `papi-shared-types.ts` that declares a pseudomodule `@papi/core` that exports all the types extensions would be expected to use. Then extensions would import like so:

```typescript
import type { ExecutionActivationContext } from '@papi/core';
```

We need to do the following:

- [x] Create new `@papi/core` as described above
- [x] Change the `.eslintrc.js` rule `import/no-unresolved` to ignore imports starting with `@papi` instead of just `papi` (in `paranext-core/extensions` and in all extension repos) and update `paranext-core/src/declarations/README.md` accordingly
- [x] In the same `.eslintrc.js` files, add [`no-restricted-imports`](https://eslint.org/docs/latest/rules/no-restricted-imports) to prevent imports from `shared/*`, `renderer/*`, `extension-host/*`, `node/*`, `client/*`, and `main/*` with error message "Importing from <whatever (like "shared")> is not allowed. Try importing from @papi/core" and a comment above that explains that these are prevented because we want people to import from `@papi/core`, not from these other modules to prevent unnecessary import breaks
- [x] Our special fake imports are as follows and need to be changed (in renderer's `global-this.model.ts` and `extension.service.ts`):
  - [x] `papi-frontend` -> `@papi/frontend`
  - [x] `papi-frontend/react` -> `@papi/frontend/react`
  - [x] `papi-backend` -> `@papi/backend`
  - [ ] `papi-shared-types` -> `@papi/shared`
- [ ] Update all the extension code to use these new imports

**Comment by tjcouch-sil** [comment]
Rolf and I discovered that there doesn't seem to be a way to unify core and bundled extension code to share types if we change `papi-shared-types` to something that can't be a file name like `@papi/shared`. Renaming the file `@papi.ts` worked for `papi.d.ts`, but then the core code would be different than extension code. So we decided to leave it as `papi-shared-types` for now.

Note: we also tried `@papi/shared/index.ts` and even creating a package.json for it, and that didn't work either.
Note: whatever you do, it needs to stay a `.ts` instead of a `.d.ts` so the code itself gets bundled into `papi.d.ts` instead of a reference made.

**Comment by tjcouch-sil** [comment]
After Rolf's changes [up to this point](https://github.com/paranext/paranext-core/compare/main...45878f9f6c01450760e864bafacacc785c148525), `use-data.hook.ts`'s errors were no longer showing. So we removed the `@ts-expect-error` lines. However, the errors would show up again when running `npm test` or `npx tsc` in `lib/papi-dts`, and the errors were still in `papi.d.ts` becuase `@ts-expect-error` was indeed finding the errors. When we commented out the following lines from `papi.d.ts` in the `@papi/core` module, the errors fixed in `papi.d.ts` and `@ts-expect-error` was failing (matching what we were seeing in core):

```typescript
// export { default as IDataProvider } from 'shared/models/data-provider.interface';
```
(fixes the errors in `use-data.hook.ts`)
```typescript
//   MandatoryProjectDataType
```
(fixes the errors in `use-project-data.hook.ts`)

It seems the problem *may* be due to using different `IDataProvider` and `MandatoryProjectDataType` types in different places. For example, core uses the originals of those types, whereas extensions use duplicate ones in `papi.d.ts`. This has already eaten lots of time and feels like a waste of time right now. Maybe we can revisit another time. Outstanding questions:

1. Why did the error go away in core?
2. Why does the error still show up when running `npm test`? And `npx tsc` in `lib/papi-dts`?
3. Could we consider using `papi.d.ts`'s `IDataProvider` and `MandatoryProjectDataType` in core? Would that fix our problem?
